### PR TITLE
Encounter type field in encounter form.  Fixes #5260

### LINKED ISSUE #5260: CQM Encounter Form needs the encounter_type code selector
### Describe the problem
User can enter the proper value for the `QDM::EncounterPerformed` so that the QRDA export can include this required value for reporting.

### Describe the solution you'd like

Expose a selector in the encounter form for the already existing field in `form_encounter` like the other items in that form:

Screenshot from 2022-04-27 22-06-43

### PR DESCRIPTION
**Pull request by adunsulag**
This eventually will fix #5260.  I'm putting this out as a draft PR so people can see what I'm doing here.

Since we can't bundle certain codes like CPT4 or even SNOMED in
non-member countries we want to only install the codes property for
the Encounter Type list options if we have the right code systems
installed.  So we have added a CodeTypes event subscriber that listens
to newly installed coded types as well as the SQL Upgrade script process
to update the Encounter Type codes property.

This is WIP and is not finalized.  I've still got to add the codes the database.sql file and do some testing on the events to make sure this is all working properly.

### CODE REVIEW

**Comment by stephenwaite** [comment]
:sunglasses:

File: interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+        foreach ($activatedCodeTypes as $record) {
+            if ($record['ct_key'] == self::CODE_TYPE_CPT4) {
+                if ($this->shouldUpdateCPT4Mappings()) {
+                    $logger("Updating " . $record['ct_key'] . " Mappings");
+                    $this->updateCPT4Mappings($logger);
+                } else {
+                    $logger("Skipping Section #updateCPT4Mappings");
+                }
+            } else if ($this->isSnomedCodeType($record['ct_key'])) {
```
Comment (stephenwaite) [inline interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php:81]: I know @sjpadgett loves `elseif` instead :), it's just a minor phpcs warning

really cool solution to this code set dilemma, great job!

Comment (sjpadgett) [reply interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php:81]: for php should be elseif. for JS will be else if. Both are acceptable unless colon shortcutting 
still safety says elseif...

File: interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        $table_records = QueryUtils::fetchRecords("SHOW TABLES LIKE ?", [$tableName]);
+        if (empty($table_records)) {
+            (new SystemLogger())->debug("table for code_type does not exist", ['tableName' => $tableName]);
+        }
+        return !empty($table_records);
+    }
+
+    private function shouldUpdateCPT4Mappings()
+    {
+        if (!$this->exists_code_table('codes_cpt')) {
```
Comment (bradymiller) [inline interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php:125]: I didn't realize there was a codes_cpt table. How does that table get created?

Comment (adunsulag) [reply interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php:125]: @bradymiller It exists in the MU3 demos and I thought that is where the CPT4 codes are loaded from.  I don't exactly know how it gets imported as the standard_tables_capture.inc doesn't have a process for it.

Comment (bradymiller) [reply interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php:125]: @adunsulag , @stephenwaite , i think most users import these into the codes table. Thoughts here? Should we check if each one exists in codes table and if it does, then ok to use the CPT code?

Comment (adunsulag) [reply interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php:125]: I just checked my codes table with the MU3 database and it has identical CPT4 codes as far as I can tell.  I think I'll just change the code up to use the codes table.  I'll make the exists_code_table to be a is_active_code_type and just check against the code_type table.

I'm glad you caught this as I was just following the same logic as what we do with snomed's sct tables.

Comment (adunsulag) [reply interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php:125]: I updated this to use the codes table.  I tested this on both the MU3 database and the demo database (openemr-cmd drid) and it looks like it is working correctly.

Comment (bradymiller) [reply interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php:125]: nice! then can bring it on in :)

**Pull request merged**